contracts/src/xchain/OmniPortal.sol

```diff
@@ -292,8 +294,18 @@ contract OmniPortal is
             // TODO: we can remove xmsg sourceChainId, and instead set _xmsg.sourceChainId to xsub.blockHeader.sourceChainId
             require(xmsg_.sourceChainId == xsub.blockHeader.sourceChainId, "OmniPortal: wrong sourceChainId");
 
-            // require xmsg conf level matches block header conf level
-            require(uint8(xmsg_.shardId) == xsub.blockHeader.confLevel, "OmniPortal: wrong confLevel");
+            if (uint8(xmsg_.shardId) == ConfLevel.Finalized) {
+                // if xmsg is Finalized, block header must be Finalized
+                require(ConfLevel.Finalized == xsub.blockHeader.confLevel, "OmniPortal: wrong conf level");
+            } else {
+                // else, require block header either matches xmsg conf level, or is finalized
+                // this allows "fuzzy" conf levels to be corrected by finalized blocks
+                require(
+                    xsub.blockHeader.confLevel == uint8(xmsg_.shardId)
+                        || xsub.blockHeader.confLevel == ConfLevel.Finalized,
+                    "OmniPortal: wrong conf level"
+                );
+            }
```

Comment:
This is core logic change

Reply:
Why do we need the `else` clause in the above statement? If the `xmsg._shardId != ConfLevel.Finalized` then any block header conf level is allowed, right? Is is added in case we support more than two confirmation levels in future?